src/Hl7.Fhir.Base/Model/IScopedNode.cs

```diff
@@ -0,0 +1,255 @@
+    /// markdown        string
+    /// base64Binary    string (uuencoded)
+    /// xhtml           string
+    /// </remarks>
+    new object? Value { get; }
+
+    /// <summary>
+    /// An indication of the location of this node within the data represented by the <c>ITypedElement</c>.
+    /// </summary>
+    /// <remarks>The format of the location is the dotted name of the property, including indices to make
+    /// sure repeated occurrences of an element can be distinguished. It needs to be sufficiently precise to aid 
+    /// the user in locating issues in the data.</remarks>
+    new string Location { get; }
+
+    /// <summary>
+    /// Resolves a reference from this node to another resource. This node should be a Bundle.
+    /// </summary>
+    /// <param name="fullUrl"></param>
+    /// <param name="result"></param>
+    /// <returns></returns>
+    public bool TryResolveBundleEntry(string fullUrl, [NotNullWhen(true)] out IScopedNode? result);
+    
+    /// <summary>
+    /// Resolves a reference from this node to another resource. This node should be a DomainResource.
+    /// </summary>
+    /// <param name="id"></param>
+    /// <param name="result"></param>
+    /// <returns></returns>
+    public bool TryResolveContainedEntry(string id, [NotNullWhen(true)] out IScopedNode? result);
+
+    /// <summary>
+    /// Resolve a resource reference within the context of this node given a url (for bundles) or id (for contained).
+    /// </summary>
+    /// <param name="url">The relative URL to resolve.</param>
+    /// <param name="result">Contains the referenced instance, or null if the operation failed</param>
+    /// <remarks>Does not create a copy. The resolved resource will be part of the IScopedNode-tree that was passed to this function</remarks>
+    /// <returns>t</returns>
+    public bool TryResolveLocalReference(string url, [NotNullWhen(true)] out IScopedNode? result)
+    {
+        for(var scan = this; scan is not null; scan = scan.Parent)
+        {
+            if (scan.Type.HasFlag(NodeType.Bundle)) // if we do not find it in the closest bundle, the reference is invalid
+            {
+                return scan.TryResolveBundleEntry(url, out result);
+            }
+            
+            if (scan.Type.HasFlag(NodeType.DomainResource) && scan.TryResolveContainedEntry(url, out result)) 
+            {
+                // if we encounter a DomainResource, try to resolve the contained reference.
+                // If it fails, higher domain resources could still contain it!
+                return true;
+            }
+
+            if (scan.Children("id").FirstOrDefault()?.Value as string == url[1..])
+            {
+                // if we encounter a resource with the correct id, return it
+                result = scan;
+                return true;
+            }
+        }
+        
+        result = null;
+        return false;
+    }
+}
+
+public static class ScopedNodeHelpers
```

Comment:
I feel this should be in a separate file, in fact we had that already (ScopedNodeExtensions)